core/domain/html_cleaner.py

```diff
@@ -143,3 +144,229 @@ def get_rte_components(html_string):
             component['customization_args'] = customization_args
             components.append(component)
     return components
+
+
+def wrap_with_siblings(tag, p):
+    """This function wraps a tag and its unwrapped sibling in p tag.
+    Args:
+        tag: bs4.element.Tag. The tag which is to be wrapped in p tag
+            along with its unwrapped siblings.
+        p: bs4.element.Tag. The new p tag in soup in which the tag and
+            its siblings are to be wrapped.
+    """
+    independent_parents = ['p', 'pre', 'ol', 'ul', 'blockquote']
+    prev_sib = list(tag.previous_siblings)
+    next_sib = list(tag.next_siblings)
+    index_of_first_unwrapped_sibling = -1
+    # Previous siblings are stored in order with the closest one
+    # being the first. All the continuous siblings which cannot be
+    # a valid parent by their own have to be wrapped in same p tag.
+    # This loop finds the index of first sibling which is a valid
+    # parent on its own.
+    for index, sib in enumerate(prev_sib):
+        if sib.name in independent_parents:
+            index_of_first_unwrapped_sibling = len(prev_sib) - index
+            break
+
+    # Previous siblings are accessed in reversed order to
+    # avoid reversing the order of siblings on being wrapped.
+    for index, sib in enumerate(reversed(prev_sib)):
+        if index >= index_of_first_unwrapped_sibling:
+            sib.wrap(p)
+
+    # Wrap the tag in same p tag as previous siblings.
+    tag.wrap(p)
+
+    # To wrap the next siblings which are not valid parents on
+    # their own in the same p tag as previous siblings.
+    for sib in next_sib:
+        if sib.name not in independent_parents:
+            sib.wrap(p)
+        else:
+            break
+
+
+def convert_to_text_angular(html_data):
```

Comment:
Unit tests? A couple of simple ones will do.

Reply:
Added!